### ISSUE TITLE
`storage_changes_notification_stream` doesn't seem to work properly

### ISSUE THREAD

**Issue by tdelabro** [paritytech/polkadot-sdk#1989]
Hey,
I'm trying to use [storage_changes_notification_stream](https://paritytech.github.io/polkadot-sdk/master/sc_client_api/client/trait.BlockchainEvents.html#tymethod.storage_changes_notification_stream).

First thing, there is no documentation or example whatsoever. The only indication we have is that `None` will act as a wildcard.

My goal is to listen to any change in a specific storage map of my pallet. If the value under any key is updated, I want a notification.

I call
```rust
storage_changes_notification_stream(
  None,
  Some(&[
    ([twox_128(MY_PALLET_NAME), twox_128(MY_STORAGE_NAME)].concat(), None)]]))
```

But it looks like because the first argument is `None` I get all the storage changes and my second arg is ignored.
I get
```
StorageChangeSet {
  changes: [<a bunch of storage changes>],
  child_changes: [], 
  filter: None,
  child_filters: Some({StorageKey([19, 67, 5, 109, 125, 191, 20, 108, 100, 100, 215, 242, 163, 70, 163, 175, 239, 212, 57, 76, 203, 26, 39, 142, 224, 180, 92, 201, 69, 43, 155, 93]): None}) }
```

So, I tried the following:
```rust
storage_changes_notification_stream(
  Some(&[]),
  Some(&[
    ([twox_128(MY_PALLET_NAME), twox_128(MY_STORAGE_NAME)].concat(), None)]]))
```

And I get nothing, no value is ever pushed to the stream.

I don't think I'm messing with my const, they are used on other parts of the node to directly access the storage without going through the runtime.
So, either I'm wrong on the way to pass an argument to this method (which is really possible given there is no documentation or examples), or the `child_filter` part just straight up does not work.
Wdyt?

**Comment by bkchr** [comment]
Are you using a child trie? If not, you need to pass your filter to `filter_keys`.

**Comment by tdelabro** [comment]
Hey @bkchr,

I will need more explanation. From my reading of the code and doc here is what is possible:
- you can pass fully fleshed `StorageKey` to `filters` in order to listen to a specific storage cell
- you can pass the beginning of a `StorageKey` (like pallet + storage but not the key of the map) to `child_filters` so you can listen to any storage change that happens below this node in the storage, and you can add a list of additional optional `StorageKey` (instead of `None` that would act as a wildcard) that would filter again the specific branch of your storage you want to listen at

Because no doc mention the way those arguments interact, I assumed they were additive.
Nowhere in the doc was mentioned something about child trie. Given there was no specific instruction about the context it should be used, I assumed it could be used in any context.

Can you please explain to me precisely how this method works, what are those arguments for, how they interact with each other, and in which context it can be used?

And hopefully, add it to the doc later on.

**Comment by bkchr** [comment]
> * you can pass the beginning of a `StorageKey` (like pallet + storage but not the key of the map) to `child_filters` so you can listen to any storage change that happens below this node in the storage

No. In Substrate there is the main trie and multiple child tries. A child trie is for example used for contracts etc. It is an independent trie outside of the main trie. Normal pallets don't use child tries. So, when you want to filter for your pallet key, you need to pass it to the first parameter of this function.

**Comment by bkchr** [comment]
https://docs.substrate.io/learn/state-transitions-and-storage/#trie-abstraction maybe this helps.

**Comment by tdelabro** [comment]
Yes did. However, it doesn't explain how these function parameters should be used.

**Comment by bkchr** [comment]
I mean yes, we should update the docs. I'm not arguing against that. I just wanted to help you.

```
storage_changes_notification_stream(
  Some(&[
    ([twox_128(MY_PALLET_NAME), twox_128(MY_STORAGE_NAME)].concat(), None)]]),
    Some(&[]),
)
```

Probably for now should work.
